ext/search_kit/Civi/Api4/Action/SearchDisplay/SaveFile.php

```diff
@@ -0,0 +1,184 @@
+  protected function processResult(\Civi\Api4\Result\SearchDisplayRunResult $result) {
+    $sk_download = \Civi\Api4\SearchDisplay::download(FALSE)
+      ->setSavedSearch($this->savedSearch)
+      ->setFormat($this->format)
+      ->setDownloadAsFile(FALSE);
+
+    if (!empty($this->seed)) {
+      $sk_download->setSeed($this->seed);
+    }
+
+    if (!empty($this->afform)) {
+      $sk_download->setAfform($this->afform);
+    }
+
+    if (!empty($this->sort)) {
+      $sk_download->setSort($this->sort);
+    }
+
+    if (!empty($this->filters)) {
+      $sk_download->setFilters($this->filters);
+    }
+
+    if (!empty($this->filterLabels)) {
+      $sk_download->setFilterLabels($this->filterLabels);
+    }
+
+    $entity_table = 'civicrm_saved_search';
+    $entity_id = $this->savedSearch['id'];
+    if (!empty($this->display['id'])) {
+      $entity_id = $this->display['id'];
+      $entity_table = 'civicrm_search_display';
+      $sk_download->setDisplay($this->display);
+    }
+
+    if ($this->appendDate) {
+      $this->fileName .= date("_Ymd", time());
+    }
+    $this->fileName .= '.' . ('array' === $this->format ? 'json' : $this->format);
+
+    // Download file contents.
+    if ('array' === $this->format) {
+      $content = $sk_download->execute();
+      $searchDisplayContent = json_encode($content);
+    }
+    else {
+      ob_start();
+      $sk_download->execute();
+      $searchDisplayContent = ob_get_contents();
+      ob_end_clean();
+    }
+
+    if (!empty($searchDisplayContent)) {
+      $config = \CRM_Core_Config::singleton();
+      $directoryName = $config->customFileUploadDir;
+      $fileName = \CRM_Utils_File::makeFileName($this->fileName);
+      $info = pathinfo($fileName);
+      // Regex in makeFileName blocks json due to check for js file names. Need to fix for saving as json.
+      if ('unknown' === $info['extension'] && 'array' === $this->format) {
+        $fileName = str_replace('unknown', 'json', $fileName);
+        $fileName = str_replace('_json', '', $fileName);
+      }
+
+      // Append folder name if provided.
+      if (!empty($this->folderName)) {
+        $folderName = preg_replace('/[-\s]/', '_', preg_replace('/[^\w\s_]/', '', $this->folderName));
+        // After replacement make sure we still have something for the folder name.
+        if (!empty($folderName)) {
+          $directoryName .= $folderName . '/';
+        }
+      }
+      \CRM_Utils_File::createDir($directoryName);
+      $fileParams = [
+        'name' => $this->fileName,
+        'mime_type' => $this->formats[$this->format]['mime'],
+        'uri' => $fileName,
+        'document' => $searchDisplayContent,
+        'description' => $this->reportName,
+        'upload_date' => date('Y-m-d H:i:s', time()),
+      ];
+
+      $fileDao = \CRM_Core_BAO_File::writeRecord($fileParams);
+      $fileDao->find(TRUE);
+
+      $entityFileDao = new \CRM_Core_DAO_EntityFile();
+      $entityFile['entity_table'] = $entity_table;
```

Comment:
There's a whitelist for this:
https://github.com/civicrm/civicrm-core/blob/26043383ebefe34e5c5b04a6d05b3141bacd9899/CRM/Core/BAO/File.php#L849-L855